File: library/src/main/java/com/alamkanak/weekview/EventChipsProvider.java
package com.alamkanak.weekview;

import android.view.View;

import org.threeten.bp.LocalDate;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import static com.alamkanak.weekview.Preconditions.checkNotNull;
import static com.alamkanak.weekview.Preconditions.checkState;

class EventChipsProvider<T> {

    private final WeekViewConfigWrapper config;
    private final WeekViewCache<T> cache;
    private final WeekViewViewState viewState;

    private WeekViewLoader<T> weekViewLoader;

    EventChipsProvider(WeekViewConfigWrapper config,
                       WeekViewCache<T> cache, WeekViewViewState viewState) {
        this.config = config;
        this.cache = cache;
        this.viewState = viewState;
    }

    void setWeekViewLoader(WeekViewLoader<T> weekViewLoader) {
        this.weekViewLoader = weekViewLoader;
    }

    void loadEventsIfNecessary(View view) {
        if (view.isInEditMode()) {
            return;
        }

        checkState(weekViewLoader != null, "No WeekViewLoader or MonthChangeListener provided.");
        final boolean hasNoEvents = cache.getAllEventChips().isEmpty();

        final LocalDate firstVisibleDay = checkNotNull(viewState.getFirstVisibleDay());
        final FetchPeriods fetchPeriods = FetchPeriods.create(firstVisibleDay);

        if (hasNoEvents || !cache.covers(fetchPeriods)) {
            loadEventsAndCalculateEventChipPositions(view, fetchPeriods);
            viewState.setShouldRefreshEvents(false);
        }
    }

    /**
     * Gets more events of one/more month(s) if necessary. This method is called when the user is
     * scrolling the week view. The week view stores the events of three months: the visible month,
     * the previous month, the next month.
     *
     * @param fetchPeriods The FetchedPeriods for which to load events
     */
    private void loadEventsAndCalculateEventChipPositions(View view, FetchPeriods fetchPeriods) {
        if (!view.isInEditMode()) {
            checkState(weekViewLoader != null, "No WeekViewLoader or MonthChangeListener provided.");
        }

        if (viewState.getShouldRefreshEvents()) {
            cache.clear();
        }

        if (weekViewLoader != null) {
            loadEvents(fetchPeriods);
        }

        calculateEventChipPositions();
    }

    private void loadEvents(FetchPeriods fetchPeriods) {
        FetchPeriods oldFetchPeriods = cache.getFetchedPeriods();
        if (oldFetchPeriods == null) {
            oldFetchPeriods = fetchPeriods;
        }

        final Period newCurrentPeriod = fetchPeriods.getCurrent();

        List<WeekViewEvent<T>> previousPeriodEvents = null;
        List<WeekViewEvent<T>> currentPeriodEvents = null;
        List<WeekViewEvent<T>> nextPeriodEvents = null;

        if (cache.getHasEvents()) {
            if (oldFetchPeriods.getPrevious().equals(newCurrentPeriod)) {
                currentPeriodEvents = cache.getPreviousPeriodEvents();
                nextPeriodEvents = cache.getCurrentPeriodEvents();
            } else if (oldFetchPeriods.getCurrent().equals(newCurrentPeriod)) {
                previousPeriodEvents = cache.getPreviousPeriodEvents();
                currentPeriodEvents = cache.getCurrentPeriodEvents();
                nextPeriodEvents = cache.getNextPeriodEvents();
            } else if (oldFetchPeriods.getNext().equals(newCurrentPeriod)) {
                previousPeriodEvents = cache.getCurrentPeriodEvents();
                currentPeriodEvents = cache.getNextPeriodEvents();
            }
        }

        if (previousPeriodEvents == null) {
            previousPeriodEvents = weekViewLoader.onLoad(fetchPeriods.getPrevious());
        }

        if (currentPeriodEvents == null) {
            currentPeriodEvents = weekViewLoader.onLoad(fetchPeriods.getCurrent());
        }

        if (nextPeriodEvents == null) {
            nextPeriodEvents = weekViewLoader.onLoad(fetchPeriods.getNext());
        }

        // Clear events.
        cache.getAllEventChips().clear();
        cache.sortAndCacheEvents(previousPeriodEvents);
        cache.sortAndCacheEvents(currentPeriodEvents);
        cache.sortAndCacheEvents(nextPeriodEvents);

        cache.setPreviousPeriodEvents(previousPeriodEvents);
        cache.setCurrentPeriodEvents(currentPeriodEvents);
        cache.setNextPeriodEvents(nextPeriodEvents);
        cache.setFetchedPeriods(fetchPeriods);
    }

    private void calculateEventChipPositions() {
        // Prepare to calculate positions of each events.
        final List<EventChip<T>> tempEvents = cache.getAllEventChips();
        final List<EventChip<T>> results = new ArrayList<>();

        // Iterate through each day with events to calculate the position of the events.
        while (!tempEvents.isEmpty()) {
            final List<EventChip<T>> eventChips = new ArrayList<>();

            final EventChip<T> firstRect = tempEvents.remove(0);
            eventChips.add(firstRect);

            int i = 0;
            while (i < tempEvents.size()) {
                // Collect all other events for same day.
                final EventChip<T> eventChip = tempEvents.get(i);
                final WeekViewEvent<T> event = eventChip.event;

                if (firstRect.event.isSameDay(event)) {
                    tempEvents.remove(i);
                    eventChips.add(eventChip);
                } else {
                    i++;
                }
            }

            computePositionOfEvents(eventChips);
            results.addAll(eventChips);
        }

        cache.put(results);
    }

    /**
     * Calculates the left and right positions of each events. This comes handy specially if events
     * are overlapping.
     *
     * @param eventChips The events along with their wrapper class.
     */
    private void computePositionOfEvents(List<EventChip<T>> eventChips) {
        // Make "collision groups" for all events that collide with others.
        final List<List<EventChip>> collisionGroups = new ArrayList<>();
        for (EventChip eventChip : eventChips) {
            boolean isPlaced = false;

            outerLoop:
            for (List<EventChip> collisionGroup : collisionGroups) {
                for (EventChip groupEvent : collisionGroup) {
                    if (groupEvent.event.collidesWith(eventChip.event)
                            && groupEvent.event.isAllDay() == eventChip.event.isAllDay()) {
                        collisionGroup.add(eventChip);
                        isPlaced = true;
                        break outerLoop;
                    }
                }
            }

            if (!isPlaced) {
                final List<EventChip> newGroup = new ArrayList<>();
                newGroup.add(eventChip);
                collisionGroups.add(newGroup);
            }
        }

        for (List<EventChip> collisionGroup : collisionGroups) {
            expandEventsToMaxWidth(collisionGroup);
        }
    }

    /**
     * Expands all the events to maximum possible width. The events will try to occupy maximum
     * space available horizontally.
     *
     * @param collisionGroup The group of events which overlap with each other.
     */
    private void expandEventsToMaxWidth(List<EventChip> collisionGroup) {
        // Expand the events to maximum possible width.
        List<List<EventChip>> columns = new ArrayList<>();
        columns.add(new ArrayList<EventChip>());

        for (EventChip eventChip : collisionGroup) {
            boolean isPlaced = false;

            for (List<EventChip> column : columns) {
                if (column.size() == 0) {
                    column.add(eventChip);
                    isPlaced = true;
                } else if (!eventChip.event.collidesWith(column.get(column.size() - 1).event)) {
                    column.add(eventChip);
                    isPlaced = true;
                    break;
                }
            }

            if (!isPlaced) {
                List<EventChip> newColumn = new ArrayList<>();
                newColumn.add(eventChip);
                columns.add(newColumn);
            }
        }

        // Calculate left and right position for all the events.
        // Get the maxRowCount by looking in all columns.
        int maxRowCount = 0;
        for (List<EventChip> column : columns) {
            maxRowCount = Math.max(maxRowCount, column.size());
        }

        for (int i = 0; i < maxRowCount; i++) {
            // Set the left and right values of the event.
            float j = 0;
            for (List<EventChip> column : columns) {
                if (column.size() >= i + 1) {
                    EventChip eventChip = column.get(i);
                    eventChip.width = 1f / columns.size();
                    eventChip.left = j / columns.size();

                    if (!eventChip.event.isAllDay()) {
                        eventChip.top = eventChip.event.getEffectiveStartMinutes(config);
                        eventChip.bottom = eventChip.event.getEffectiveEndMinutes(config);
                    } else {
                        eventChip.top = 0;
                        eventChip.bottom = 100;
                    }
                }
                j++;
            }
        }
    }

}

File: library/src/main/java/com/alamkanak/weekview/EventChipsProvider.kt
package com.alamkanak.weekview

internal class EventChipsProvider<T>(
    private val config: WeekViewConfigWrapper,
    private val cache: WeekViewCache<T>,
    private val viewState: WeekViewViewState
) {

    var monthLoader: MonthLoader<T>? = null

    fun loadEventsIfNecessary() {
        val hasNoEvents = cache.allEventChips.isEmpty()
        val shouldRefresh = viewState.shouldRefreshEvents

        val firstVisibleDay = checkNotNull(viewState.firstVisibleDay)
        val fetchPeriods = FetchPeriods.create(firstVisibleDay)

        if (hasNoEvents || shouldRefresh || !cache.covers(fetchPeriods)) {
            loadEventsAndCalculateEventChipPositions(fetchPeriods)
            viewState.shouldRefreshEvents = false
        }
    }

    private fun loadEventsAndCalculateEventChipPositions(fetchPeriods: FetchPeriods) {
        if (viewState.shouldRefreshEvents) {
            cache.clear()
        }

        loadEvents(fetchPeriods)
        calculateEventChipPositions()
    }

    private fun loadEvents(fetchPeriods: FetchPeriods) {
        val oldFetchPeriods = cache.fetchedPeriods ?: fetchPeriods
        val newCurrentPeriod = fetchPeriods.current

        var previousPeriodEvents: List<WeekViewEvent<T>>? = null
        var currentPeriodEvents: List<WeekViewEvent<T>>? = null
        var nextPeriodEvents: List<WeekViewEvent<T>>? = null

        if (cache.hasEvents) {
            when (newCurrentPeriod) {
                oldFetchPeriods.previous -> {
                    currentPeriodEvents = cache.previousPeriodEvents
                    nextPeriodEvents = cache.currentPeriodEvents
                }
                oldFetchPeriods.current -> {
                    previousPeriodEvents = cache.previousPeriodEvents
                    currentPeriodEvents = cache.currentPeriodEvents
                    nextPeriodEvents = cache.nextPeriodEvents
                }
                oldFetchPeriods.next -> {
                    previousPeriodEvents = cache.currentPeriodEvents
                    currentPeriodEvents = cache.nextPeriodEvents
                }
            }
        }

        val loader = checkNotNull(monthLoader) { "No WeekViewLoader or MonthChangeListener provided." }

        if (previousPeriodEvents == null) {
            previousPeriodEvents = loader.load(fetchPeriods.previous)
        }

        if (currentPeriodEvents == null) {
            currentPeriodEvents = loader.load(fetchPeriods.current)
        }

        if (nextPeriodEvents == null) {
            nextPeriodEvents = loader.load(fetchPeriods.next)
        }

        cache.update(previousPeriodEvents, currentPeriodEvents, nextPeriodEvents, fetchPeriods)
    }

    private fun calculateEventChipPositions() {
        val results = mutableListOf<EventChip<T>>()
        val groups = cache.allEventChips.groupBy { it.event.startTime.atStartOfDay }

        for (eventChips in groups.values) {
            computePositionOfEvents(eventChips)
            results += eventChips
        }

        cache.put(results)
    }

    /**
     * Forms [CollisionGroup]s for all event chips and uses them to expand the [EventChip]s to their
     * maximum width.
     *
     * @param eventChips A list of [EventChip]s
     */
    private fun computePositionOfEvents(eventChips: List<EventChip<T>>) {
        val collisionGroups = mutableListOf<CollisionGroup<T>>()

        for (eventChip in eventChips) {
            val collidingGroup = collisionGroups.firstOrNull { it.collidesWith(eventChip) }

            if (collidingGroup != null) {
                collidingGroup.add(eventChip)
            } else {
                collisionGroups += CollisionGroup(eventChip)
            }
        }

        for (collisionGroup in collisionGroups) {
            expandEventsToMaxWidth(collisionGroup.eventChips)
        }
    }

    /**
     * Expand all [EventChip]s in a [CollisionGroup] to their maximum width.
     *
     * @param eventChips The [EventChip]s of a [CollisionGroup]
     */
    private fun expandEventsToMaxWidth(eventChips: List<EventChip<T>>) {
        val columns = mutableListOf<Column<T>>(Column())

        for (eventChip in eventChips) {
            val fittingColumn = columns.firstOrNull { it.fits(eventChip) }
            if (fittingColumn != null) {
                fittingColumn.add(eventChip)
            } else {
                columns += Column(eventChip)
            }
        }

        val rows = columns.map { it.size }.max() ?: 0

        for (row in 0 until rows) {
            columns.forEachIndexed { index, column ->
                val hasEventInRow = column.size > row

                if (hasEventInRow) {
                    val eventChip = column[row]

                    // Every column gets the same width. For instance, if there are four columns,
                    // then each column's width is 0.25.
                    eventChip.width = 1f / columns.size

                    // The start position is calculated based on the index of the column. For
                    // instance, if there are four columns, the start positions will be 0, 0.25, 0.5
                    // and 0.75.
                    eventChip.left = index.toFloat() / columns.size

                    eventChip.calculateTopAndBottom(config)
                }
            }
        }
    }

    /**
     * This class encapsulates [EventChip]s that collide with each other, meaning that they overlap
     * from a time perspective.
     *
     */
    private class CollisionGroup<T>(
        val eventChips: MutableList<EventChip<T>>
    ) {

        constructor(eventChip: EventChip<T>) : this(mutableListOf(eventChip))

        /**
         * Returns whether an [EventChip] collides with any [EventChip] already in the
         * [CollisionGroup].
         *
         * @param eventChip An [EventChip]
         * @return Whether a collision exists
         */
        fun collidesWith(eventChip: EventChip<T>): Boolean {
            return eventChips.any { it.event.collidesWith(eventChip.event) }
        }

        fun add(eventChip: EventChip<T>) {
            eventChips.add(eventChip)
        }

    }

    /**
     * This class encapsulates [EventChip]s that are displayed in the same column.
     */
    private class Column<T>(
        val eventChips: MutableList<EventChip<T>> = mutableListOf()
    ) {

        constructor(eventChip: EventChip<T>) : this(mutableListOf(eventChip))

        val isEmpty: Boolean
            get() = eventChips.isEmpty()

        val size: Int
            get() = eventChips.size

        fun add(eventChip: EventChip<T>) {
            eventChips.add(eventChip)
        }

        operator fun get(index: Int): EventChip<T> {
            return eventChips[index]
        }

        fun fits(eventChip: EventChip<T>): Boolean {
            return isEmpty || !eventChips.last().event.collidesWith(eventChip.event)
        }

    }

}
